### PR TITLE
v4.0.x: reproducible builds + portability fix

### PR DESCRIPTION
**Pull request by jsquyres**
Cherry pick of #8136 and #8152 from master regarding reproducible builds.  Thanks to @bmwiedemann for bringing this back to our attention.

Also bring in a MacOS/`readlink` portability fix.

### CODE REVIEW

**Comment by bmwiedemann** [comment]
This and 4.1 needs one more patch:
```diff
orte/tools/orte-info/Makefile.am
-            -DOMPI_BUILD_DATE="\"`date`\"" \
+            -DOMPI_BUILD_DATE="\"`$(top_srcdir)/config/getdate.sh`\"" \
```

**Comment by jsquyres** [comment]
@bmwiedemann I notice on master that in one place, we have:

```sh
${USER:-`whoami`}
```

but in other places we just have

```sh
$$USER
```

Should they all be of the form 

```sh
# With $ or $$, as appropriate
{USER:-`whoami`}
```

?

**Comment by bmwiedemann** [comment]
I'd leave it as is. I have seen good PRs with feature-creep in the past that got stalled too long.

**Comment by jsquyres** [comment]
Ah, I see -- `$USER` is an automatically-defined shell variable, anyway (I thought it was something unique to Reproducible builds, and therefore might not actually be set).

**Comment by rhc54** [comment]
Is it safe to say we need similar changes in PMIx and PRRTE?

**Comment by bmwiedemann** [comment]
@rhc54 I tested with our `openmpi4` package and these changes (with orte-info) were enough to make it reproducible.

**Comment by rhc54** [comment]
Understood - my apologies. My question was more about building PMIx and PRRTE standalone packages (i.e., when they aren't embedded in OMPI)

**Comment by jsquyres** [comment]
@rhc54 It would not be a bad thing to also apply this to PMIx or PRRTE.  I wouldn't say it's critical, but it's definitely in the "nice to have" category.

**Comment by jsquyres** [comment]
This PR will also need the MacOS portability fixes from #8152.

**Comment by jsquyres** [comment]
bot:ompi:retest

**Review by rhc54 — APPROVED** [review]

File: orte/tools/orte-info/Makefile.am [hunk 1]
```diff
@@ -26,7 +26,7 @@ AM_CFLAGS = \
             -DOPAL_CONFIGURE_DATE="\"@OPAL_CONFIGURE_DATE@\"" \
             -DOMPI_BUILD_USER="\"$$USER\"" \
             -DOMPI_BUILD_HOST="\"`(hostname || uname -n) | sed 1q`\"" \
```
Comment (bmwiedemann) [inline orte/tools/orte-info/Makefile.am:28]: Tested it and this line still caused variations.
=>
```
\"$${HOSTNAME:-`(hostname || uname -n) 2> /dev/null | sed 1q`}\"" \
```

Comment (jsquyres) [reply orte/tools/orte-info/Makefile.am:28]: oshmem_info and mpisync will therefore cause the same issue.

Comment (rhc54) [reply orte/tools/orte-info/Makefile.am:28]: So we need that line to change? FWIW: that suggested line is what we have in PRRTE today (someone asked for reproducible build over there earlier)

Comment (jsquyres) [reply orte/tools/orte-info/Makefile.am]: I just pushed that change here (and #8148) for orte-info and mpisync.  Anywhere you have a `-DBUILD_HOST` kind of construct in PMIX/PRTE Makefile.am's, you'll want to allow the `$HOSTNAME` override.

**Review by bmwiedemann — APPROVED** [review]

**Pull request merged**